### DISCUSSION TITLE
ActiveFS questions

### DISCUSSION THREAD

**Discussion by hrstoyanov** [activej/activej#187] [Q&A]
Thank you for releasing ActiveJ family of products!
I am trying to assess ActiveFS for a distributed file system that can run under a database server and provide seamless database replication. I am comparing Active FS possible solution to [Aeron](https://github.com/real-logic/aeron) + [Dokan Java](https://github.com/dokan-dev/dokan-java) implementation and have these questions:

1. Does ActveFS support low level block operations (like update Nth block in an existing remote file). This is essential for building distributed file systems. It seems ActiveFS supports only very limited high-level file operations (commands)

2. How does ActiveFS coordinate the the consistent updates to a distributed file? Aeron provides Raft protocol clustering where there is always a Master and Replicas , and new master can be elected in case of failure. How is ActiveFS P2P doing this?

3. The ActiveFS examples only show single client-server interactions. Are there example for cluster-wide operations? 

4. Aeron is extremely efficient by utilizing UDP level broadcasting in the cluster.  It goes to extreme care to avoid Java GC overhead and use of "lock-free" structures to avoid threading issues. How does ActiveFS compare - HTTP is not very efficient transmission protocol and not good for broadcasting?

**Comment by eduard-vasinskyi** [comment]
Hello, @hrstoyanov

ActiveFS is a thin abstraction on top of Java's NIO channels. It provides methods to asynchronously work with files using ActiveJ's promises and CSP. As a result, the overhead of ActiveFs is really low. 

There are two main implementations:
- **LocalActiveFs** for processing files on a local filesystem
- **ClusterActiveFs** for processing files stored on remote partitions

ClusterActiveFs was designed to serve as an optimal storage for processing big data. A common task for big data is working with immutable files. ClusterActiveFs allows you to do just that, as all files stored in a cluster are immutable. This greatly simplifies synchronization and allows us to create a system that is highly effective.
ClusterActiveFs does not have masters or replicas. It adheres to a pure peer-to-peer design which makes scaling and failover quite simple. 

There are currently no examples for ClusterActiveFs, but you can look at tests in [cluster package](https://github.com/activej/activej/tree/master/cloud-fs/src/test/java/io/activej/fs/cluster) to get an idea of how it can be used. We will be sure to update our website with proper examples for ClusterActiveFs as soon as possible.

ActiveFs supports operations via HTTP ([HttpActiveFs](https://github.com/activej/activej/blob/master/cloud-fs/src/main/java/io/activej/fs/http/HttpActiveFs.java)/[ActiveFsServlet](https://github.com/activej/activej/blob/master/cloud-fs/src/main/java/io/activej/fs/http/ActiveFsServlet.java)) and raw TCP([RemoteActiveFs](https://github.com/activej/activej/blob/master/cloud-fs/src/main/java/io/activej/fs/tcp/RemoteActiveFs.java) / [ActiveFsServer](https://github.com/activej/activej/blob/master/cloud-fs/src/main/java/io/activej/fs/tcp/ActiveFsServer.java)).

**Comment by valerialistratova** [comment]
Hello, @hrstoyanov. We've updated ActiveFS docs to cover the FAQs about this component. Please consider taking a look - https://fs.activej.io

**Comment by hrstoyanov** [comment]
Thank you, it is clearer now!
